src/multiboot-headers.md

```diff
@@ -282,7 +282,21 @@ Again, we’ll use math to let the computer calculate the sum for us. We add up
 the magic number, the mode code, and the header length, and then subtract it
 from a big number. `dd` then puts that value into this spot in our file.
 
-TODO: https://github.com/intermezzOS/book/issues/28
+> By the way...
+>
+> 0xe85250d6 is 3,897,708,758 in decimal, and the smallest number you can store as a 32-bit signed integer is -2,147,483,648, so we get an integer underflow. This doesn't matter for the checksum, but since underflows are usually a bug the assembler gives us a warning.
```

Comment:
Huh, looking into this is I just realized that the original description of what is happening here is mostly wrong. The more I looked into the original math (which I'd taken for granted), it didn't seem like an underflow or overflow would be happening at all. I ran the code for myself without the 0x100000000, and I wasn't able to coerce any warning or errors out of the nasm assembler, just a failed boot which looked like a bad checksum.

Turns out [the multiboot spec](http://nongnu.askapache.com/grub/phcoder/multiboot.pdf) says nothing at all about modulo operators. It says:

> The field `checksum` is a 32-bit unsigned value which, when added to the other magic fields (i.e. ‘magic’, ‘architecture’ and ‘header_length’), must have a 32-bit unsigned sum of zero.

In other words:

`checksum` + `magic_number` + `architecture` + `header_length` = 0

We could try and "solve for" `checksum` like so:

`checksum` =  -(`magic_number` + `architecture` + `header_length`)

But of course we can't represent -(`magic_number` + `architecture` + `header_length`) in a uint32, due to all the wraparound issues. So instead we come up with a number that "tricks" the ALU into giving us zero when we do the addition. Imagine for the sake of argument that `magic_number` + `architecture` + `header_length` somehow works out to be 0xFFFFFFFE. The number we'd add to that in order to make 0 would be 0x00000002. This is 0x100000000-0xFFFFFFFE, because 0x100000000 technically maps to 0 with all the wraparound math.

(On top of all that, the checksum isn't even a signed integer, which makes sense, there's no point in signing things at this low a level!.)

So I guess I got a bunch of really good practice explaining what a signed integer is in my local branch here. I'll rewrite this entire description to be accurate.